chain/chain/src/store.rs

```diff
@@ -1733,111 +1765,212 @@ impl<'a> ChainStoreUpdate<'a> {
-                // 1b. Delete incoming receipts (ColIncomingReceipts)
-                store_update
-                    .delete(ColIncomingReceipts, &get_block_shard_id(&block_hash, shard_id));
-                self.chain_store
-                    .incoming_receipts
-                    .cache_remove(&get_block_shard_id(&block_hash, shard_id));
-                // 1c. Delete from chunk_hash_per_height_shard (ColChunkPerHeightShard)
-                store_update.delete(ColChunkPerHeightShard, &get_height_shard_id(height, shard_id));
-                self.chain_store
-                    .chunk_hash_per_height_shard
-                    .cache_remove(&get_height_shard_id(height, shard_id));
-                // 1d. Delete from next_block_with_new_chunk (ColNextBlockWithNewChunk)
-                store_update
-                    .delete(ColNextBlockWithNewChunk, &get_block_shard_id(&block_hash, shard_id));
-                self.chain_store
-                    .next_block_with_new_chunk
-                    .cache_remove(&get_block_shard_id(&block_hash, shard_id));
-            }
-            for chunk_header in block.chunks {
-                let (receipts, transactions) = match self.get_chunk_clone_from_header(&chunk_header)
-                {
-                    Ok(chunk) => (chunk.receipts, chunk.transactions),
-                    _ => (vec![], vec![]),
-                };
-                // 1e. Delete from receipt_id_to_shard_id (ColReceiptIdToShardId)
-                for receipt in receipts {
-                    store_update.delete(ColReceiptIdToShardId, receipt.receipt_id.as_ref());
-                    self.chain_store
-                        .receipt_id_to_shard_id
-                        .cache_remove(&receipt.receipt_id.into());
-                }
-                // 1f. Delete from ColTransactions
-                for transaction in transactions {
-                    store_update.delete(ColTransactions, transaction.get_hash().as_ref());
-                    self.chain_store.transactions.cache_remove(&transaction.get_hash().into());
+        if let Ok(canonical_hash) = self.get_block_hash_by_height(height) {
+            for block_hash in blocks_current_height.iter() {
+                if *block_hash != canonical_hash {
+                    // Block `block_hash` is not on the Canonical Chain and may be safely deleted
+                    // and all its ancestors while there are no other sibling blocks rely on it.
+                    let mut current_hash = *block_hash;
+                    loop {
+                        if self.get_block_refcount(&current_hash) == 0 {
+                            let prev_hash = self.get_block_header(&current_hash)?.prev_hash;
+
+                            self.clear_block_data(trie.clone(), current_hash, true)?;
+
+                            current_hash = prev_hash;
+                            if current_hash == self.get_block_hash_by_height(height)? {
```

Comment:
need to use height of block with `current_hash`

Reply:
++